spec/fastsync/version2/scheduler.tla

```diff
@@ -0,0 +1,580 @@
+                    
+\* the state of the scheduler:
+VARIABLES outEvent,       \* an event from the scheduler to the environment
+          scRunning
+
+\* the block pool: 
+VARIABLE scheduler
+    (*
+       scheduler is a record that contains: 
+         height: Int,
+            height of the next block to collect
+         peers: PeerIDs,
+            the set of peers that have connected in the past, may include removed peers
+         peerHeights: [PeerIDs -> Heights],
+            a map to collect the peer heights, >0 if peer in ready state, None (-1) otherwise
+         peerStates: [PeerIDs -> PeerStates],
+            a map to record the peer states 
+         blockStates: [Heights -> BlockStates]
+            a set of heights for which blocks are to be scheduled, pending or received
+         pendingBlocks: [Heights -> PeerIDs],
+            a set of heights for which blocks are to be scheduled or pending
+         receivedBlocks: [Heights -> PeerIDs],
+            a set of heights for which blocks were received but not yet processed
+    *)
+    
+vars == <<turn, envRunning, inEvent, scRunning, outEvent, scheduler>>    
+
+\* for now just keep the height in the block
+Blocks == [ height: Heights ]
+
+noEvent == [type |-> "NoEvent"] 
+
+InEvents == 
+  {noEvent} \cup
+  [type: {"rTrySchedule", "tNoAdvanceExp", "rTryPrunePeer"}] \cup
+  [type: {"bcStatusResponse"},    peerID: PeerIDs, height: Heights] \cup
+  [type: {"bcBlockResponse"},     peerID: PeerIDs, height: Heights, block: Blocks] \cup
+  [type: {"bcNoBlockResponse"},   peerID: PeerIDs, height: Heights] \cup
+  [type: {"pcBlockProcessed"},    peerID: PeerIDs, height: Heights] \cup
+  [type: {"pcBlockVerificationFailure"}, firstPeerID: PeerIDs, secondPeerID: PeerIDs] \cup
+  [type: {"bcAddNewPeer"},        peerID: PeerIDs] \cup
+  [type: {"bcRemovePeer"},        peerID: PeerIDs]
+
+\* Output events produced by the scheduler.
+\* Note: in v2 the status request is done by the reactor/ environment
+OutEvents == 
+  {noEvent} \cup
+  [type: {"scPeerError"}, peerID: PeerIDs, reason: Errors] \cup
+  [type: {"scSchedulerFail"}, reason: Errors] \cup
+  [type: {"scBlockRequest"}, peerID: PeerIDs, height: Heights] \cup
+  [type: {"scBlockReceived"}, peerID: PeerIDs, block: Blocks] \cup
+  [type: {"scPeersPruned"}, pruned: SUBSET [peerID: PeerIDs]] \cup 
+  [type: {"scFinishedEv"},  reason: Errors]
+                        
+(* ----------------------------------------------------------------------------------------------*)
+(* The behavior of the scheduler that keeps track of peers, block requests and responses, etc.   *)
+(* See scheduler.go                                                                              *)
+(* https://github.com/tendermint/tendermint/tree/brapse/blockchain-v2-riri-reactor-2/scheduler.go*)
+(* ----------------------------------------------------------------------------------------------*)
+
+addPeer(sc, peerID) == 
```

Comment:
It feels that (at least in case of Golang) we can automatically generate TLA+ model of the code that would be pretty close to this version.